src/runtime_strings_extra.js

```diff
@@ -117,16 +119,17 @@ function lengthBytesUTF16(str) {
   return str.length*2;
 }
 
-function UTF32ToString(ptr) {
+function UTF32ToString(ptr, maxCharsToRead) {
 #if ASSERTIONS
   assert(ptr % 4 == 0, 'Pointer passed to UTF32ToString must be aligned to four bytes!');
 #endif
   var i = 0;
+  if (maxCharsToRead === undefined) maxCharsToRead = Infinity;
 
   var str = '';
   while (1) {
     var utf32 = {{{ makeGetValue('ptr', 'i*4', 'i32') }}};
-    if (utf32 == 0) return str;
+    if (utf32 == 0 || i >= maxCharsToRead) return str;
```

Comment:
Instead of having a `while(1)` with `i >= maxCharsToRead`, it would be better to move the index check to the `while`.

These changes to UTF16ToString and UTF32ToString regress code size. I am still keeping my hopes up that Closure compiler would one day be able to optimize it (https://github.com/google/closure-compiler/issues/3193, https://github.com/google/closure-compiler/issues/3201 and https://github.com/google/closure-compiler/issues/3223). Could you change this to use the same compare against NaN/undefined trick that UTF8ToString uses to save code size and make the code (perhaps) optimizable by Closure in the future?